File: argon-options/src/main/scala/dev/argon/options/OptionDecoder.scala
package dev.argon.options

import esexpr.{ESExpr, ESExprCodec}
import zio.*
import esexpr.ESExprCodec
import dev.argon.io.ResourceReader
import java.io.IOException

trait OptionDecoder[A] {
  def decode(expr: ESExpr): ZIO[ResourceReader, ESExprCodec.DecodeError, A]
}

File: plugin/api/java/src/main/java/dev/argon/plugin/api/options/OptionDecoder.java
package dev.argon.plugin.api.options;

import dev.argon.plugin.api.ResourceFactory;
import dev.argon.plugin.api.tube.Toml;
import org.jetbrains.annotations.NotNull;

import java.util.Optional;

public interface OptionDecoder<E extends Throwable, A> {
    @NotNull A decode(@NotNull ResourceFactory<E> resourceFactory, @NotNull Toml value) throws OptionDecodeException;
    default @NotNull Optional<A> defaultValue() {
        return Optional.empty();
    }
}
